src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -592,8 +593,18 @@ internal void UpdateCurrentSolutionOnOptionsChanged()
 #pragma warning disable IDE0060 // Remove unused parameter
     protected internal Task ScheduleTask(Action action, string? taskName = "Workspace.Task")
     {
-        _workQueue.AddWork(action);
-        return _workQueue.WaitUntilCurrentBatchCompletesAsync();
+        // Require main thread on the callback as this is publicly exposed and the given actions may have main thread dependencies.
+        var handlerAndOptions = new WorkspaceEventHandlerAndOptions(args => action(), WorkspaceEventOptions.RequiresMainThreadOptions);
+        var handlerSet = EventHandlerSet.Create(handlerAndOptions);
+
+        return ScheduleTask(EventArgs.Empty, handlerSet);
+    }
+
+    [SuppressMessage("Style", "VSTHRD200:Use \"Async\" suffix for async methods", Justification = "This is a Task wrapper, not an asynchronous method.")]
+    internal Task ScheduleTask(EventArgs args, EventHandlerSet handlerSet)
```

Comment:
Maybe call this AddToEventHandlerWorkQueue or something, so it's very clear this isn't doing the thread scheduling that the existing method is doing.

Reply:
I think I'm going to avoid this one in this PR. All the "ScheduleTask" methods go through this method, so they are all kind of misnamed now. Add to that the whole Task returning behaviors for these methods, and I think this needs to be thought through a bit more.